demos/mqtt/subscription-manager/mqtt_subscription_manager.c

```diff
@@ -0,0 +1,448 @@
+ * @param[in] topicFilterLength Length of topic filter.
+ *
+ * @return `true` if the topic name and topic filter match; `false`
+ * otherwise.
+ */
+static bool matchTopic( const char * pTopicName,
+                        const uint16_t topicNameLength,
+                        const char * pTopicFilter,
+                        const uint16_t topicFilterLength );
+
+/*-----------------------------------------------------------*/
+
+static bool matchEndWildcards( const char * pTopicFilter,
+                               uint16_t topicFilterLength,
+                               uint16_t topicNameLength,
+                               uint16_t nameIndex,
+                               uint16_t filterIndex )
+{
+    bool matchFound = false, endChar = false;
+
+    /* Determine if the last character is reached for the topic name, and the
+     * second last character for the topic filter to match for the '#' wildcard. */
+    endChar = ( nameIndex == ( topicNameLength - 1U ) ) &&
+              ( filterIndex == ( topicFilterLength - 3U ) ) ? true : false;
+
+    if( endChar == true )
+    {
+        /* Determine if the topic filter contains "/#" as the last 2 characters for the '#' wildcard.
+         * The '#' wildcard represents the parent and any number of child levels in the topic name.
+         * For example, the filter "sport/#" matches "sport" as well as "sport/tennis" topics. */
+        matchFound = ( ( pTopicFilter[ filterIndex + 1U ] == '/' ) &&
+                       ( pTopicFilter[ filterIndex + 2U ] == '#' ) ) ? true : false;
+    }
+
+    if( matchFound == false )
+    {
+        /* Determine if the last character is reached for both topic name and topic
+         * filter to match for the '+' wildcard. */
+        endChar = ( nameIndex == ( topicNameLength - 1U ) ) &&
+                  ( filterIndex == ( topicFilterLength - 2U ) ) ? true : false;
+
+        if( endChar == true )
+        {
+            /* Filter "sport/+" matches the "sport/" but not "sport". */
+            matchFound = ( pTopicFilter[ filterIndex + 1U ] == '+' ) ? true : false;
+        }
+    }
+
+    return matchFound;
+}
+
+/*-----------------------------------------------------------*/
+
+static bool matchWildcards( const char * pTopicFilter,
+                            const char * pTopicName,
+                            uint16_t topicNameLength,
+                            uint16_t filterIndex,
+                            uint16_t * pNameIndex,
+                            bool * pMatch )
+{
+    bool shouldStopMatching = false;
+
+    /* Check for wildcards. */
+    if( pTopicFilter[ filterIndex ] == '+' )
+    {
+        /* Move topic name index to the end of the current level.
+         * This is identified by '/'. */
+        while( ( *pNameIndex < topicNameLength ) && ( pTopicName[ *pNameIndex ] != '/' ) )
+        {
+            ( *pNameIndex )++;
+        }
+
+        ( *pNameIndex )--;
```

Comment:
Please write a comment why this move back is needed.

- If the break condition is `( *pNameIndex < topicNameLength )`, the corresponding index increment is done in `topicFilterMatch`.
- If the break condition is `( pTopicName[ *pNameIndex ] != '/' )`, we move back the index on the '/' character. 

In the second case, do we match the '/' character in the topicName and filter or both the indexes are just skipped in the `topicFilterMatch`? In other words, would filter `/topic/+abc` match `/topic/part/bc`?

Reply:
Thanks for the suggestion. 

For your question on:
> In the second case, do we match the '/' character in the topicName and filter or both the indexes are just skipped in the topicFilterMatch? In other words, would filter /topic/+abc match /topic/part/bc?

The `topicFilterMatch` will match the `/` character in the next iteration of the loop after both the nameIndex and filterIndex indices are incremented, thus the invalid topic filter `/topic/+abc` cannot match `/topic/part/bc`